### DISCUSSION TITLE
v2 separate image update automation per stage

### DISCUSSION THREAD

**Discussion by donotaskwho** [fluxcd/flux2#2591] [Q&A] [answered]
Hello!

We would like to flux to be updating images per specific stage. Previously in v1, we were able to create different images for each stage and flux would update them accordingly using flux-patch. In v2, we're not clear on what to use in order to achieve the same result. 

Currently, when the latest timestamped image tag has been uploaded to the registry, flux detects it and updates the deployment which pushes the new change to all stages. This is not something we want as we'd like to keep manual approval for production systems. 

What can be used in each stage for this purpose? Tagging each stage image is not a problem, more about how to configure each stage detecting only the image tag with the corresponding name.

Any ideas, examples, or guides would be greatly appreciated!

Thanks!

**Comment by kingdonb** [comment, marked as answer]
If you are using different branches for each stage, include the branch name in your image tags when you design your tagging strategy according to [the Sortable Image Tags guide](https://fluxcd.io/docs/guides/sortable-image-tags/)

You can create a filter for each stage in the form of an `ImagePolicy` so that only images built from the `dev` branch are deployed in `dev`, and only `stage` images in `stage` for example, I don't know what your environments are called, but for example anyway...

Another possibility is to use semver ranges, so that staging environment can be defined with a policy like `1.x.x-0` which will deploy any prerelease build or release build. That way, when your final release is deployed in production, you don't have to go back and update staging separately.

Hope this helps!

**Reply by donotaskwho** [reply]
Thanks for the reply! 

I'd like to try the `ImagePolicy` with filters for each stage (we're using different image tags per stage, not branches). 
Correct me if I'm wrong but in that case, each service should have a separate ImagePolicy stored in `stages/<specific-stage>` or is there a way to define it once? _(we're using the recommended folder structure)_

**Reply by kingdonb** [reply]
The ImagePolicy, ImageUpdateAutomation, and ImageRepository do not care what namespace the resources that they target are in, the `ImageUpdateAutomation` only cares about the path, as it operates on resources in git and not in the cluster.

The primary consideration for where these resources should be placed then, is related to the secret from `sourceRef` in IUA. There must be a secret with read-write access to the repository, as described in the guide. This is obviously a sensitive credential, as it can lead to privilege escalation and other very bad outcomes if it was compromised.

Long story short, they do not 100% need to be located together in the same namespace anymore, but if you are using soft multi-tenancy then you will have disabled cross-namespace references according to https://github.com/fluxcd/flux2-multi-tenancy and then they will need to be co-located together in the same namespace.

If your tenants manage their own image update policies, then they should go together in a tenant namespace. If your system is centrally managed and policies should be set by Flux admins, then they can be placed together in the flux-system namespace. In the multi-tenant case, they should go wherever the GitRepository goes, because they cannot reach across namespaces to access a secret.

**Reply by donotaskwho** [reply]
The problem is not in the namespaces it is in the different contexts. Our stages are in different Kubernetes contexts and we would like `imagepolicy` to be specific to each of these contexts. Can you advise how to do that? 

There is no cross-namespace management or stage-specific branches. We'd only like for the image policy to pick up the regex pattern differently for each of the context.

**Reply by kingdonb** [reply]
Each image policy selects a single tag as the latest for that policy.

You could create multiple policies, one for each context. The policy only contains one rule to determine which images pass the filter and which sorting approach should be used (numeric, alphabetic).

So you would create an ImagePolicy named `staging` for the staging context where (say, for example) any release with a prerelease version can be deployed:

```
  semver: 1.x.x-0
```

Then another `production` which does not deploy prereleases:

```
  semver: 1.x.x
```

and another `dev` that deploys the latest build [based on a timestamp](https://fluxcd.io/docs/guides/sortable-image-tags/#using-in-an-imagepolicy-object)...

```
  filterTags:
    pattern: '^.*-[a-f0-9]+-(?P<ts>[0-9]+)$'
    extract: '$ts'
```

you can unify these under a single ImageUpdateAutomation for the namespace. Each policy only selects one latest tag, and only works on one ImageRepository at a time. This makes for an unfortunate proliferation of these automation resources required to automate many different apps (but this is the current state of how it works.)

Then in each context (say we are in the namespace "podinfo" and we have three contexts here) you mark up the manifest with the policy that should be used there.

```
# {"$imagepolicy": "podinfo:production"}
# {"$imagepolicy": "podinfo:staging"}
# {"$imagepolicy": "podinfo:dev"}
```

What else I have learned since April is that, ImageUpdateAutomation and ImagePolicy must be in the same namespace. Even if cross-namespace references are allowed. I do not understand the reason for this limitation as of yet, since the ImagePolicy kyaml setter directive contains the namespace, I think it would be possible to reach across since this markup is quite explicit.

I am working on some answers for how to practically apply this pattern without manually repeating yourself and managing all this boilerplate by hand. For now this is how it works. If there's more that I can do to help, please join the CNCF slack and ask in #flux. It might be easier to get the right idea in a synchronous setting, in case there's more about your structure I did not quite understand.
